tests/unit_tests/test_datacube.py

```diff
@@ -0,0 +1,218 @@
+    assert output["data"]["data"] == expected_data
+    assert output["data"]["shape"] == expected_shape
+
+
+@pytest.mark.parametrize(
+    "shape,null_axes,expected_coords",
+    [
+        ([2, 2], [], [[0, 0], [0, 1], [1, 0], [1, 1]]),
+        ([1], [], [[0]]),
+        ([3], [], [[0], [1], [2]]),
+        ([2, 2, 2], [], [[0, 0, 0], [0, 0, 1], [0, 1, 0], [0, 1, 1], [1, 0, 0], [1, 0, 1], [1, 1, 0], [1, 1, 1]]),
+        ([2, 2], [0], [[None, 0], [None, 1]]),
+        ([2, 2], [1], [[0, None], [1, None]]),
+        ([2, 2], [0, 1], [[None, None]]),
+        ([2, 2, 2], [0, 1, 2], [[None, None, None]]),
+        ([2, 2, 2], [0, 1], [[None, None, 0], [None, None, 1]]),
+        ([2, 2, 2], [0, 2], [[None, 0, None], [None, 1, None]]),
+        ([2, 2, 2], [1], [[0, None, 0], [0, None, 1], [1, None, 0], [1, None, 1]]),
+        ([2, 2, 2], [2], [[0, 0, None], [0, 1, None], [1, 0, None], [1, 1, None]]),
+    ],
+)
+def test_iterateCoords(datacube_code, shape, null_axes, expected_coords):
+    testing_code = (
+        datacube_code([])
+        + f"\nfor(let c of datacube._iterateCoords({json.dumps(shape)}, {json.dumps(null_axes)})) {{console.log(c)}}"
+    )
+    output = run_javascript(testing_code).decode("utf-8")
+    coords = output.strip().split("\n")
+
+    for i, coord in enumerate(coords):
+        coord = json.loads(coord)
+        assert coord in expected_coords
+        expected_coords.remove(coord)
+
+    assert len(expected_coords) == 0
+
+
+@pytest.mark.parametrize(
+    "example_data,expected_data_shape,dimensions,dimension_to_reduce,expected_data,expected_shape",
+    [
+        ([1, 2, 3, 4], [2, 2], None, "b", [3, 7], [2]),
+        ([1, 2, 3, 4], [2, 2], None, "t", [4, 6], [2]),
+        (
+            [1, 2, 3, 4, 5, 6, 7, 8],
+            [2, 2, 2],
+            [
+                {"name": "x", "labels": [], "type": "other"},
+                {"name": "t", "labels": [], "type": "temporal"},
+                {"name": "b", "labels": [], "type": "bands"},
+            ],
+            "b",
+            [3, 7, 11, 15],
+            [2, 2],
+        ),
+        (
+            [1, 2, 3, 4, 5, 6, 7, 8],
+            [2, 2, 2],
+            [
+                {"name": "x", "labels": [], "type": "other"},
+                {"name": "t", "labels": [], "type": "temporal"},
+                {"name": "b", "labels": [], "type": "bands"},
+            ],
+            "t",
+            [4, 6, 12, 14],
+            [2, 2],
+        ),
+        (
+            [1, 2, 3, 4, 5, 6, 7, 8],
+            [2, 2, 2],
+            [
+                {"name": "x", "labels": [], "type": "other"},
+                {"name": "t", "labels": [], "type": "temporal"},
+                {"name": "b", "labels": [], "type": "bands"},
+            ],
+            "x",
+            [6, 8, 10, 12],
+            [2, 2],
+        ),
+        ([1, 2], [2], [{"name": "t", "labels": [], "type": "temporal"}], "t", [3], None),
+    ],
+)
+def test_reduceByDimension(
```

Comment:
These tests somewhat cover similar cases as the ones in [function unit tests](https://github.com/openEOPlatform/openeo-pg-evalscript-converter/blob/master/tests/unit_tests/test_reduce_dimension.py) . It's ok to have both, but I believe the ones here should be more fine grained and the former should be more general (and we should have less of them).

Reply:
Well, the tests here call the `reduceByDimension` directly (instead of testing the process) and also tests more general examples (more than 2 dimensions etc.). Imo it doesn't hurt to have extra tests?

Reply:
> Well, the tests here call the reduceByDimension directly (instead of testing the process) and also tests more general examples (more than 2 dimensions etc.).

This is cool - that we test more here. All test cases covering `reduceByDimension` functionality should reside here. In the process we just need to make sure that `reduce_dimension` is working from the process - we don't need to cover all the cases again there since that functionality has already been tested.

> Imo it doesn't hurt to have extra tests?

We don't want to have duplicate tests or tests cases covering the same scenarios. It will be slower to run, more work to maintain, harder to debug etc.

Reply:
I actually meant we should remove tests in `test_reduce_dimension` as they are actually the ones that duplicate tests in data cube `reduceByDimension`. (`reduceByDimension` is more atomic).

But it's fine as it is, let's not overcomplicate this. fftr.